File: 0997-find-the-town-judge/0997-find-the-town-judge.js
/**
 * @param {number} n
 * @param {number[][]} trust
 * @return {number}
 */
var findJudge = function (n, trust) {
    if (n === 1 && trust.length === 0) return 1;

    const outdegree = new Map();
    const indegree = new Map();

    for (let i = 0; i < trust.length; i++) {
        let pair = trust[i];

        outdegree.set(pair[0], (outdegree.get(pair[0]) ?? 0) + 1);
        if (!outdegree.has(pair[1])) {
            outdegree.set(pair[1], 0);
        }
        indegree.set(pair[1], (indegree.get(pair[1]) ?? 0) + 1);
    }

    for (let [key, value] of outdegree) {
        if (value === 0 && indegree.get(key) === n - 1) {
            return key;
        }
    }

    return -1;
};

File: 0997-find-the-town-judge/0997-find-the-town-judge.py
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        if n == 1 and len(trust) == 0:
            return 1

        indegree = {}
        outdegree = {}

        for i in range(len(trust)):
            pair = trust[i]

            outdegree[pair[0]] = outdegree.get(pair[0], 0) + 1
            if pair[1] not in outdegree:
                outdegree[pair[1]] = 0
            indegree[pair[1]] = indegree.get(pair[1], 0) + 1

        for key, value in outdegree.items():
            if value == 0 and indegree.get(key, 0) == n - 1:
                return key

        return -1
